enterprise/trackedusers/data_collector.go

```diff
@@ -0,0 +1,192 @@
+package trackedusers
+
+//go:generate mockgen -destination=./mocks/mock_data_collector.go -package=mockdatacollector github.com/rudderlabs/rudder-server/enterprise/trackedusers DataCollector
+
+import (
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"time"
+
+	migrator "github.com/rudderlabs/rudder-server/services/sql-migrator"
+
+	obskit "github.com/rudderlabs/rudder-observability-kit/go/labels"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+
+	"github.com/rudderlabs/rudder-go-kit/logger"
+
+	"github.com/lib/pq"
+
+	"github.com/spaolacci/murmur3"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/segmentio/go-hll"
+
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	txn "github.com/rudderlabs/rudder-server/utils/tx"
+)
+
+const (
+	idTypeUserID                       = "userID"
+	idTypeAnonymousID                  = "anonymousID"
+	idTypeUserIDAnonymousIDCombination = "userIDAnonymousIDCombination"
+
+	// changing this will be non backwards compatible
+	murmurSeed = 123
+)
+
+// DataCollector is interface to collect data from jobs
+type DataCollector interface {
+	CollectData(ctx context.Context, jobs []*jobsdb.JobT, tx *txn.Tx) error
+}
+
+type UniqueUsersCollector struct {
+	log         logger.Logger
+	hllSettings *hll.Settings
+	instanceID  string
+}
+
+func NewUniqueUsersCollector(log logger.Logger, dbConn string) (*UniqueUsersCollector, error) {
+	dbHandle, err := sql.Open("postgres", dbConn)
+	if err != nil {
+		panic(err)
+	}
+	dbHandle.SetMaxOpenConns(1)
+
+	m := &migrator.Migrator{
+		Handle:                     dbHandle,
+		MigrationsTable:            "tracked_users_reports_migrations",
+		ShouldForceSetLowerVersion: config.GetBool("SQLMigrator.forceSetLowerVersion", true),
+	}
+	err = m.Migrate("tracked_users")
+	if err != nil {
+		return nil, fmt.Errorf("could not run tracked_users_reports migrations: %w", err)
+	}
+
+	return &UniqueUsersCollector{
+		log: log,
+		hllSettings: &hll.Settings{
+			Log2m:             config.GetInt("TrackedUsers.precision", 14),
+			Regwidth:          config.GetInt("TrackedUsers.registerWidth", 5),
+			ExplicitThreshold: hll.AutoExplicitThreshold,
+			SparseEnabled:     true,
+		},
+		instanceID: config.GetString("INSTANCE_ID", "1"),
+	}, nil
+}
+
+func (u *UniqueUsersCollector) CollectData(ctx context.Context, jobs []*jobsdb.JobT, tx *txn.Tx) error {
+	if len(jobs) == 0 {
+		return nil
+	}
+	workspaceSourceUserIdTypeMap := make(map[string]map[string]map[string]*hll.Hll)
+	reportedAt := time.Now().UTC()
+	for _, job := range jobs {
+		if job.WorkspaceId == "" {
+			u.log.Warn("workspace_id not found in job", logger.NewIntField("jobId", job.JobID))
+			continue
+		}
+
+		sourceID := gjson.GetBytes(job.Parameters, "source_id").String()
+		if sourceID == "" {
+			u.log.Warn("source_id not found in job parameters", obskit.WorkspaceID(job.WorkspaceId),
+				logger.NewIntField("jobId", job.JobID))
+			continue
+		}
+		userID := gjson.GetBytes(job.EventPayload, "batch.0.userId").String()
```

Comment:
Just double checking if each job corresponds to one event. and we're fetching userId at event level and not for a batch. There's a chance of a batch API call to have multiple userIds(like a backend service doing batch API calls). I'm assuming even if we store in gw jobs under batch, we store one event only. rt?
Also, we should skip non event stream sources(retl, etl etc)

Reply:
each job can have multiple events but events are batched per user. Adding filter for non event stream sources